File: rayforge/core/geo/cleanup.py
from __future__ import annotations
import numpy as np
from typing import List, Tuple, Optional, TYPE_CHECKING

from .constants import (
    CMD_TYPE_MOVE,
    CMD_TYPE_LINE,
    CMD_TYPE_ARC,
    CMD_TYPE_BEZIER,
    COL_TYPE,
    COL_X,
    COL_Y,
    COL_Z,
    COL_I,
    COL_J,
    COL_CW,
    COL_C1X,
    COL_C1Y,
    COL_C2X,
    COL_C2Y,
)


if TYPE_CHECKING:
    from .geometry import Geometry


def _are_points_equal(
    p1: Tuple[float, ...],
    p2: Tuple[float, ...],
    tolerance: float,
) -> bool:
    """Check if two points are equal within tolerance."""
    if len(p1) != len(p2):
        return False
    for i in range(len(p1)):
        if abs(p1[i] - p2[i]) > tolerance:
            return False
    return True


def _get_segment_key(
    data: np.ndarray,
    idx: int,
    tolerance: float,
) -> Optional[Tuple]:
    """Get a hashable key for a segment at the given index."""
    if idx >= len(data):
        return None

    row = data[idx]
    cmd_type = row[COL_TYPE]

    if cmd_type == CMD_TYPE_LINE:
        end_point = (row[COL_X], row[COL_Y], row[COL_Z])
        return ("LINE", end_point)

    if cmd_type == CMD_TYPE_ARC:
        end_point = (row[COL_X], row[COL_Y], row[COL_Z])
        center_offset = (row[COL_I], row[COL_J])
        clockwise = bool(row[COL_CW])
        return ("ARC", end_point, center_offset, clockwise)

    if cmd_type == CMD_TYPE_BEZIER:
        end_point = (row[COL_X], row[COL_Y], row[COL_Z])
        c1 = (row[COL_C1X], row[COL_C1Y])
        c2 = (row[COL_C2X], row[COL_C2Y])
        return ("BEZIER", end_point, c1, c2)

    return None


def _are_segments_equal(
    key1: Tuple,
    key2: Tuple,
    tolerance: float,
) -> bool:
    """Check if two segment keys represent identical segments."""
    if key1[0] != key2[0]:
        return False

    seg_type = key1[0]

    if seg_type == "LINE":
        return _are_points_equal(key1[1], key2[1], tolerance)

    if seg_type == "ARC":
        return (
            _are_points_equal(key1[1], key2[1], tolerance)
            and _are_points_equal(key1[2], key2[2], tolerance)
            and key1[3] == key2[3]
        )

    if seg_type == "BEZIER":
        return (
            _are_points_equal(key1[1], key2[1], tolerance)
            and _are_points_equal(key1[2], key2[2], tolerance)
            and _are_points_equal(key1[3], key2[3], tolerance)
        )

    return False


def remove_duplicate_segments(
    data: Optional[np.ndarray],
    tolerance: float = 1e-6,
) -> Optional[np.ndarray]:
    """
    Remove duplicate segments from geometry data.

    This function identifies and removes duplicate line, arc, and bezier
    segments. A segment is considered duplicate if it has the same type,
    endpoint, and parameters as another segment within the same path.

    Move commands are preserved as they define path boundaries and reset
    duplicate detection for new paths.

    Parameters
    ----------
    data : np.ndarray
        Geometry data array of shape (N, 8) where each row represents
        a command. Columns are [type, x, y, z, param1, param2, param3,
        param4].
    tolerance : float, optional
        Maximum distance for two points to be considered equal.
        Default is 1e-6.

    Returns
    -------
    np.ndarray
        Geometry data array with duplicate segments removed.

    Examples
    --------
    >>> import numpy as np
    >>> from rayforge.core.geo.constants import CMD_TYPE_MOVE, CMD_TYPE_LINE
    >>> data = np.array([
    ...     [CMD_TYPE_MOVE, 0, 0, 0, 0, 0, 0, 0],
    ...     [CMD_TYPE_LINE, 10, 0, 0, 0, 0, 0, 0],
    ...     [CMD_TYPE_LINE, 10, 0, 0, 0, 0, 0, 0],
    ... ])
    >>> result = remove_duplicate_segments(data)
    >>> len(result)
    2
    """
    if data is None or len(data) == 0:
        return data

    keep_mask = np.ones(len(data), dtype=bool)
    seen_segments: List[Tuple] = []

    for i in range(len(data)):
        row = data[i]
        cmd_type = row[COL_TYPE]

        if cmd_type == CMD_TYPE_MOVE:
            seen_segments = []
            continue

        key = _get_segment_key(data, i, tolerance)

        if key is None:
            continue

        is_duplicate = False
        for seen_key in seen_segments:
            if _are_segments_equal(key, seen_key, tolerance):
                is_duplicate = True
                break

        if is_duplicate:
            keep_mask[i] = False
        else:
            seen_segments.append(key)

    return data[keep_mask]


def close_geometry_gaps_from_array(
    data: np.ndarray, tolerance: float = 1e-6
) -> np.ndarray:
    """
    Closes small gaps in a geometry array to form clean, connected paths.

    Args:
        data: The input geometry numpy array.
        tolerance: The maximum distance between two points to be
                    considered "the same".

    Returns:
        A new, modified numpy array.
    """
    if data is None or len(data) < 2:
        return data if data is not None else np.array([])

    modified_data = data.copy()
    move_indices = np.where(modified_data[:, COL_TYPE] == CMD_TYPE_MOVE)[0]
    sub_arrays = np.split(modified_data, move_indices[1:])

    for sub in sub_arrays:
        if len(sub) < 2:
            continue
        start_pt = sub[0, COL_X : COL_Z + 1]
        end_pt = sub[-1, COL_X : COL_Z + 1]
        dist_sq = np.sum((start_pt - end_pt) ** 2)
        if dist_sq < tolerance * tolerance:
            sub[-1, COL_X : COL_Z + 1] = start_pt
    modified_data = np.vstack(sub_arrays)

    final_rows: List[np.ndarray] = []
    last_end_point: np.ndarray | None = None
    for row in modified_data:
        cmd_type = row[COL_TYPE]
        end_point = row[COL_X : COL_Z + 1]

        if cmd_type == CMD_TYPE_MOVE:
            if last_end_point is not None:
                dist_sq = np.sum((end_point - last_end_point) ** 2)
                if dist_sq < tolerance * tolerance:
                    new_row = row.copy()
                    new_row[COL_TYPE] = CMD_TYPE_LINE
                    new_row[COL_X : COL_Z + 1] = last_end_point
                    final_rows.append(new_row)
                else:
                    final_rows.append(row)
                    last_end_point = end_point
            else:
                final_rows.append(row)
                last_end_point = end_point
        else:
            final_rows.append(row)
            last_end_point = end_point

    if not final_rows:
        return np.array([])
    return np.array(final_rows)


def close_geometry_gaps(
    geometry: Geometry, tolerance: float = 1e-6
) -> Geometry:
    """
    Closes small gaps in a Geometry object to form clean, connected paths.

    This function creates a new Geometry object with the modifications.

    Args:
        geometry: The input Geometry object.
        tolerance: The maximum distance between two points to be
                    considered "the same".

    Returns:
        A new, modified Geometry object.
    """
    new_geo = geometry.copy()
    new_geo._sync_to_numpy()
    if new_geo.is_empty() or new_geo._data is None:
        return new_geo

    new_geo._data = close_geometry_gaps_from_array(
        new_geo._data, tolerance=tolerance
    )
    return new_geo

File: crates/rayforge-geo/src/path/cleanup.rs
//! Cleanup: Deduplication and gap closing for geometry data.
//!
//! Provides functions for cleaning geometry command arrays by removing
//! duplicate segments and closing small gaps between connected paths.

use crate::types::Command;

/// Check if two points (as 3-element arrays) are equal within a tolerance.
pub fn are_points_equal(p1: &[f64; 3], p2: &[f64; 3], tolerance: f64) -> bool {
    for i in 0..3 {
        if (p1[i] - p2[i]).abs() > tolerance {
            return false;
        }
    }
    true
}

/// Extract a hashable key for a segment at the given index.
/// Returns None for MOVE commands.
pub fn get_segment_key(row: &[f64; 8]) -> Option<(u32, [f64; 3], [f64; 4])> {
    let cmd = Command::from_row(row).ok()?;
    match cmd {
        Command::Move { .. } => None,
        Command::Line { end } => Some((2, [end.0, end.1, end.2], [0.0; 4])),
        Command::Arc {
            end,
            center_offset,
            clockwise,
        } => {
            let params = [
                center_offset.0,
                center_offset.1,
                if clockwise { 1.0 } else { 0.0 },
                0.0,
            ];
            Some((3, [end.0, end.1, end.2], params))
        }
        Command::Bezier {
            end,
            control1,
            control2,
        } => {
            let params = [control1.0, control1.1, control2.0, control2.1];
            Some((4, [end.0, end.1, end.2], params))
        }
    }
}

/// Check if two segment keys represent identical segments within tolerance.
pub fn are_segments_equal(
    k1: &(u32, [f64; 3], [f64; 4]),
    k2: &(u32, [f64; 3], [f64; 4]),
    tolerance: f64,
) -> bool {
    if k1.0 != k2.0 {
        return false;
    }
    if !are_points_equal(&k1.1, &k2.1, tolerance) {
        return false;
    }
    if k1.0 == 2 {
        return true;
    }
    if k1.0 == 3 {
        return are_points_equal(
            &[k1.2[0], k1.2[1], 0.0],
            &[k2.2[0], k2.2[1], 0.0],
            tolerance,
        ) && (k1.2[2] - k2.2[2]).abs() < tolerance;
    }
    if k1.0 == 4 {
        let p1 = [k1.2[0], k1.2[1], k1.2[2]];
        let p2 = [k2.2[0], k2.2[1], k2.2[2]];
        return are_points_equal(&p1, &p2, tolerance)
            && (k1.2[3] - k2.2[3]).abs() < tolerance;
    }
    false
}

/// Remove duplicate segments from geometry command data.
///
/// A segment is considered duplicate if it has the same type, endpoint,
/// and parameters as another segment within the same subpath. MOVE commands
/// reset duplicate detection for new paths.
pub fn remove_duplicate_segments(
    data: &[[f64; 8]],
    tolerance: f64,
) -> Vec<[f64; 8]> {
    if data.is_empty() {
        return data.to_vec();
    }

    let mut result: Vec<[f64; 8]> = Vec::new();
    let mut seen_segments: Vec<(u32, [f64; 3], [f64; 4])> = Vec::new();

    for row in data {
        let cmd = Command::from_row(row).expect("invalid command");

        if matches!(cmd, Command::Move { .. }) {
            seen_segments.clear();
            result.push(*row);
            continue;
        }

        if let Some(key) = get_segment_key(row) {
            let is_dup = seen_segments
                .iter()
                .any(|sk| are_segments_equal(&key, sk, tolerance));
            if is_dup {
                continue;
            }
            seen_segments.push(key);
        }
        result.push(*row);
    }

    result
}

/// Close small gaps in a geometry data array to form clean, connected paths.
///
/// Two operations are performed:
/// 1. Within each subpath (between MOVE commands), if the end point is within
///    tolerance of the start point, snap the end point to the start.
/// 2. If a MOVE command starts within tolerance of the previous subpath's end,
///    convert it to a LINE command (bridging the gap).
pub fn close_geometry_gaps_from_array(
    data: &[[f64; 8]],
    tolerance: f64,
) -> Vec<[f64; 8]> {
    if data.len() < 2 {
        return data.to_vec();
    }

    let tol_sq = tolerance * tolerance;

    // Find MOVE command indices to identify subpath boundaries
    let mut move_indices: Vec<usize> = Vec::new();
    for (i, row) in data.iter().enumerate() {
        let cmd = Command::from_row(row).expect("invalid command");
        if matches!(cmd, Command::Move { .. }) {
            move_indices.push(i);
        }
    }

    let mut modified: Vec<[f64; 8]> = data.to_vec();

    // Build ranges for each subpath
    let sub_ranges: Vec<(usize, usize)> = if move_indices.is_empty() {
        vec![(0, data.len())]
    } else {
        let mut ranges = Vec::new();
        let mut prev = 0;
        for &mi in &move_indices[1..] {
            ranges.push((prev, mi));
            prev = mi;
        }
        ranges.push((prev, data.len()));
        ranges
    };

    // Close gaps within each subpath
    for &(start, end) in &sub_ranges {
        if end - start >= 2 {
            let s = Command::from_row(&modified[start])
                .expect("invalid command")
                .end_point();
            let e_cmd =
                Command::from_row(&modified[end - 1]).expect("invalid command");
            let e = e_cmd.end_point();
            let dsq =
                (s.0 - e.0).powi(2) + (s.1 - e.1).powi(2) + (s.2 - e.2).powi(2);
            if dsq < tol_sq {
                let new_cmd = match e_cmd {
                    Command::Move { .. } => Command::Move { end: s },
                    Command::Line { .. } => Command::Line { end: s },
                    Command::Arc {
                        center_offset,
                        clockwise,
                        ..
                    } => Command::Arc {
                        end: s,
                        center_offset,
                        clockwise,
                    },
                    Command::Bezier {
                        control1, control2, ..
                    } => Command::Bezier {
                        end: s,
                        control1,
                        control2,
                    },
                };
                modified[end - 1] = new_cmd.to_row();
            }
        }
    }

    // Convert MOVE to LINE when at the same position as previous endpoint
    let mut final_rows: Vec<[f64; 8]> = Vec::new();
    let mut last_end: Option<(f64, f64, f64)> = None;

    for &row in &modified {
        let cmd = Command::from_row(&row).expect("invalid command");
        let end_pt = cmd.end_point();

        if matches!(cmd, Command::Move { .. }) {
            if let Some(prev) = last_end {
                let dsq = (end_pt.0 - prev.0).powi(2)
                    + (end_pt.1 - prev.1).powi(2)
                    + (end_pt.2 - prev.2).powi(2);
                if dsq < tol_sq {
                    let line = Command::Line { end: prev };
                    final_rows.push(line.to_row());
                } else {
                    final_rows.push(row);
                }
            } else {
                final_rows.push(row);
            }
        } else {
            final_rows.push(row);
        }
        last_end = Some(end_pt);
    }

    final_rows
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::constants::*;

    #[test]
    fn test_remove_duplicate_lines() {
        let data: Vec<[f64; 8]> = vec![
            [CMD_TYPE_MOVE, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            [CMD_TYPE_LINE, 10.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            [CMD_TYPE_LINE, 10.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
        ];
        let result = remove_duplicate_segments(&data, 1e-6);
        assert_eq!(result.len(), 2);
    }

    #[test]
    fn test_remove_duplicate_different() {
        let data: Vec<[f64; 8]> = vec![
            [CMD_TYPE_MOVE, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            [CMD_TYPE_LINE, 10.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            [CMD_TYPE_LINE, 10.0, 10.0, 0.0, 0.0, 0.0, 0.0, 0.0],
        ];
        let result = remove_duplicate_segments(&data, 1e-6);
        assert_eq!(result.len(), 3);
    }

    #[test]
    fn test_close_gaps_simple() {
        let data: Vec<[f64; 8]> = vec![
            [CMD_TYPE_MOVE, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            [CMD_TYPE_LINE, 10.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            [CMD_TYPE_LINE, 10.0, 10.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            [CMD_TYPE_LINE, 0.1, 0.1, 0.0, 0.0, 0.0, 0.0, 0.0],
        ];
        let result = close_geometry_gaps_from_array(&data, 0.2);
        assert_eq!(result.last().unwrap()[COL_X], 0.0);
        assert_eq!(result.last().unwrap()[COL_Y], 0.0);
    }
}
